provider/datagen/src/transform/cldr/time_zones/names.rs

```diff
@@ -0,0 +1,107 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::convert::compute_bcp47_tzids_btreemap;
+use crate::transform::cldr::cldr_serde;
+use icu_provider::datagen::IterableDataProvider;
+use icu_provider::prelude::*;
+use icu_timezone::provider::names::*;
+use icu_timezone::TimeZoneBcp47Id;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::hash::Hasher;
+use zerotrie::ZeroTriePerfectHash;
+use zerovec::ZeroVec;
+
+impl DataProvider<IanaToBcp47MapV1Marker> for crate::DatagenProvider {
+    fn load(&self, _: DataRequest) -> Result<DataResponse<IanaToBcp47MapV1Marker>, DataError> {
+        let resource: &cldr_serde::time_zones::bcp47_tzid::Resource =
+            self.cldr()?.bcp47().read_and_parse("timezone.json")?;
+
+        let iana2bcp = &compute_bcp47_tzids_btreemap(&resource.keyword.u.time_zones.values);
+
+        // Sort and deduplicate the BCP-47 IDs:
+        let bcp_set: BTreeSet<TimeZoneBcp47Id> = iana2bcp.values().copied().collect();
+        let bcp47_ids: ZeroVec<TimeZoneBcp47Id> = bcp_set.iter().copied().collect();
+
+        // Transform the map to use BCP indices:
+        #[allow(clippy::unwrap_used)] // structures are derived from each other
+        let map: BTreeMap<Vec<u8>, usize> = iana2bcp
+            .iter()
+            .map(|(iana, bcp)| {
+                (
+                    iana.as_bytes().to_ascii_lowercase().to_vec(),
+                    bcp47_ids.binary_search(bcp).unwrap(),
+                )
+            })
+            .collect();
+
+        // Compute the checksum:
+        let mut hasher = twox_hash::XxHash64::with_seed(0);
+        hasher.write(bcp47_ids.as_bytes());
```

Comment:
why `as_bytes`? we require the elements to be the same and this relies on a ZV invariant we shouldn't need to know when reading this code